values/encoding/string_encode.go

```diff
@@ -112,12 +111,26 @@ func (enc *stringEncoder) Encode(strVals values.StringValues, writer io.Writer)
 
 	switch metaProto.Encoding {
 	case encodingpb.EncodingType_RAW_SIZE:
-		return enc.rawSizeEncode(valuesIt, writer)
+		if err := enc.rawSizeEncode(valuesIt, writer); err != nil {
+			return err
+		}
 	case encodingpb.EncodingType_DICTIONARY:
-		return enc.dictionaryEncode(valuesIt, dictionary, writer)
+		if err := enc.dictionaryEncode(valuesIt, dictionary, writer); err != nil {
+			return err
+		}
 	default:
 		return fmt.Errorf("invalid encoding type: %v", metaProto.Encoding)
 	}
+
+	// Close the writer if it satisifies the `io.WriteCloser` iface.
+	wc, ok := writer.(io.WriteCloser)
```

Comment:
It's probably more efficient if you pull `compressWriter` declaration outside the `switch` block and do a nil check here than doing a type cast.